### ISSUE TITLE
All Context functions should return a Result instead of calling expect()

### ISSUE THREAD

**Issue by federicomenaquintero** [gtk-rs/cairo#338]
Things have gotten to a point in librsvg where most of the incoming bugs that make the library panic are due to cairo-rs calling `expect()` on an `Err` inside the `Context` methods.  (That is, rather than panicking in librsvg itself, something in the drawing calls causes the `cr` to enter an error state.)

For example, all of the `line_to` / `move_to` / etc. can set the `cr` to an out-of-memory state if they can't allocate the new command.  In theory, they could also set an error state if they detected coordinates outside Cairo's supported range (I intend to look into that soon - Cairo crashes really easily if given coordinates outside ±2^23, and it doesn't do range checking for coordinates).

I'd like to propose that either all the Context functions return a Result, or that we remove the expect()s and make `Context.status()` public again.

One thing librsvg would like to do, for example, is to feed a path to cairo with the `move_to`, `line_to`, etc. and just check at the end if there was an error.  It would also be fine to catch a result with `?` and return as early as possible if there is an error in one of the individual commands.

Cairo contexts and surfaces have the same "lenient" error checking - they can go into an error state and ignore further API calls, but this is just to help C programs avoid having `if (cairo_something(...) != OK)` everywhere.  We have better language facilities in Rust :smiley: 

Summary: 

* I think just returning a Result instead of calling expect() would let the calling program make the correct choice of what to do upon errors.

* I think we still want a public way to get the Context's status.  One thing librsvg does is to get a `*mut cairo_t` from the C world, then wraps it in a `Context` and checks to see if the context is in error - it will g_warning() and return if so, to avoid panicking inside librsvg/cairo-rs.

**Comment by GuillaumeGomez** [comment]
I **completely** agree and would be more in favor for returning `Result`/`Option` more often.

**Comment by federicomenaquintero** [comment]
Also, I don't think returning Result everywhere is too onerous.  Code that already looks like this:

```rust
cr.move_to(...);
cr.line_to(...);
cr.curve_to();
cr.close_path();
cr.fill();
```

Will simply turn into this:
```rust
cr.move_to(...)?;
cr.line_to(...)?;
cr.curve_to()?;
cr.close_path()?;
cr.fill()?;
```

**Comment by sdroege** [comment]
> I **completely** agree

I also agree for this specific case in cairo. Want to send a PR @federicomenaquintero ?

**Comment by sdroege** [comment]
>     * One thing librsvg does is to get a `*mut cairo_t` from the C world, then wraps it in a `Context` and checks to see if the context is in error - it will g_warning() and return if so, to avoid panicking inside librsvg/cairo-rs.

The conversion from `*mut cairo_t` to `cairo::Context` should already take care that the context is not in an error state. Maybe we should add functions for returning a `Result` for these two in addition to the currently panicking ones (whose signature is needed for fitting with the way how GLib FFI works)?

**Comment by federicomenaquintero** [comment]
> I also agree for this specific case in cairo. Want to send a PR @federicomenaquintero ?

I think I'll start a branch to sketch this and post it for comments.  For example, I'm not yet sure if "making invalid matrices panic" goes well with this; I found that *really* helpful in librsvg to find bugs, and then inconvenient after the real bugs were fixed.

**Comment by federicomenaquintero** [comment]
> Maybe we should add functions for returning a Result for these two in addition to the currently panicking ones (whose signature is needed for fitting with the way how GLib FFI works)?

Yeah, probably similar to how surfaces are done; `from_raw_full(pointer) -> Result`.

**Comment by federicomenaquintero** [comment]
Hmm, the "from a pointer" functions for `Context` are the same as for surfaces, and the `unwrap` here is making me uncomfortable:

```rust
impl FromGlibPtrFull<*mut ffi::cairo_t> for Context {
    unsafe fn from_glib_full(ptr: *mut ffi::cairo_t) -> Context {
        Self::from_raw_full(ptr).unwrap()
    }
}
```

* What if we didn't implement `FromGlibPtrFull`?
* Or, do we need a more general `trait TryFromGlibPtrFull`?  Do "pointer conversions that can fail" happen elsewhere in the gtk-rs universe?

**Comment by sdroege** [comment]
> What if we didn't implement `FromGlibPtrFull`?

Then various things in gdk/gtk couldn't be autogenerated anymore. Also arguably if some C API gives you an invalid context/surface, exploding is the best thing you can do at that point. If that happens then everything's lost anyway.

> Or, do we need a more general `trait TryFromGlibPtrFull`? Do "pointer conversions that can fail" happen elsewhere in the gtk-rs universe?

gobject-introspection doesn't have any concept of such thing, so that would be hard :)

Why are you looking at those specific parts though? That shouldn't be needed for making the API inside cairo better or am I missing something?

**Comment by federicomenaquintero** [comment]
Work in progress here: https://github.com/federicomenaquintero/cairo/commits/results-everywhere

> Then various things in gdk/gtk couldn't be autogenerated anymore. Also arguably if some C API gives you an invalid context/surface, exploding is the best thing you can do at that point. If that happens then everything's lost anyway.

Hmmmm, can you point me to that autogenerated stuff?  Just to check my assumptions on cairo-rs's guarantees.

I'm not sure if librsvg is the only thing that checks about getting passed a cairo_t in error from the C world; I definitely found it useful to implement a "nope, I can't work with that crap" kind of check for that :)  Usually the problem was that people were creating surfaces larger than Cairo supports, so it would give back an error surface - people would create a cairo_t from it, which is in an error state, and feed that to librsvg to paint to the surface.

**Comment by sdroege** [comment]
> Hmmmm, can you point me to that autogenerated stuff? Just to check my assumptions on cairo-rs's guarantees.

https://github.com/gtk-rs/gtk/blob/9b9cf0f623bd74bf71459f639beac2f6f26c163e/src/auto/widget.rs#L3731-L3765

If GTK gives us a broken context there then that's a big problem. I'm not sure what else than panicking would make sense there.

> I definitely found it useful to implement a "nope, I can't work with that crap" kind of check for that :)

That's the `g_return_val_if_fail()` / `panic!()` case though. If the caller passes in garbage then that's a programming error, and returning a `Result` is not going to make that go away. `Result`s are for expected errors

**Comment by sdroege** [comment]
@federicomenaquintero Any updates here?

**Comment by federicomenaquintero** [comment]
Sorry, I'll get back to it this week.

(Do we have any release schedule / should I just try to stick to GNOME's?)

**Comment by sdroege** [comment]
It's ready whenever it's ready :)
